### PR TITLE
fix(chat): apply the Fast capability gate to the native passthrough (#1886)

### PR DESCRIPTION
**Pull request by olddonkey**
## Summary

The last known bug from the FastWire umbrella (#1886): the native `/v1/chat/completions` passthrough decided service-tier forwarding with one provider-level check and bypassed the policy layer entirely.

```ts
if (provider.chatServiceTier && rawBody.service_tier !== undefined) {
  body.service_tier = rawBody.service_tier;
}
```

**The worst consequence is not the one A0 recorded.** A0 locked this as "exact-model `false` is ignored", but the same line also ignored an explicit `supportsServiceTier: false` — the fail-closed declaration that means *this upstream is documented not to accept the field*. Whenever `chatServiceTier` happened to be true, a caller tier still reached that upstream. B1's split between capability-authorized canonical Fast and CallerTierForward-authorized foreign tiers never reached this surface either.

## The fix

Both Chat surfaces now call **one shared gate**, so they cannot drift apart by construction rather than by convention — a stronger guarantee than parallel implementations kept honest by tests. The passthrough resolves its policy in `chat-native.ts`, where the route and config already live, and the key-rotation rebuild path resolves it too.

| Configuration | Before | After |
| --- | --- | --- |
| `supportsServiceTier: false` + `chatServiceTier: true`, caller sends a tier | forwarded | stripped (fail-closed restored) |
| exact-model `false`, caller sends a tier | forwarded | stripped |
| exact-model `true`, no `chatServiceTier`, caller sends canonical Fast | forwarded | forwarded (capability authorizes it) |
| exact-model `true`, no `chatServiceTier`, caller sends `flex` | forwarded | stripped (foreign tiers need CallerTierForward) |
| unclassified, no `chatServiceTier` | forwarded | stripped |

## One difference that stays, on purpose

The Responses surface normalizes a caller's `"fast"` spelling to the provider's canonical wire value; this passthrough forwards the caller's exact bytes. That is this builder's documented contract — *caller fields retain their exact wire representation while capability gates stay centralized* — and the two halves map cleanly onto this change: the gate is now shared, the bytes are still the caller's.

The distinction is defensible beyond the doc comment: on the Responses path a `"fast"` spelling originates in Codex's own configuration vocabulary, so normalizing it is right; on this surface an OpenAI-compatible client chose that literal value. A test asserts the divergence and the call site explains it, so a future reader does not mistake it for a missed translation.

## Known gap, deliberately left for its own unit

The passthrough neither runs through the adapter registry decorator nor calls `recordAdapterTier`, so it produces no `AttemptTierOutcome` — a stripped tier is not yet visible as `callerTierDropped` in the logs. Building half of that pipeline inside a bug fix would have been worse than naming it here.

## Tests

- The A0 characterization cell for this bug is flipped and annotated, per the protocol #1904 established. Every other A0 cell passes untouched.
- New: both fail-closed levels, canonical-vs-foreign authorization separation, unclassified with and without the opt-in, and a case asserting `"FAST"` survives verbatim on this surface.
- New: a parity block asserting both Chat paths reach the *same forward/strip decision* for identical configuration and input.
- `tests/openrouter-provider-routing.test.ts` updated for the signature change; its assertions are unchanged.

## Verification

- Full suite, best of three runs: **13,332 pass / 10 skip / 2 fail**.
- One failure is constant across every run: `tests/key-login-live-update.test.ts`, the pre-existing dev-side regression flagged on #1893/#1904/#1956/#2072, which reproduces on pristine `dev`.
- The rest were timeouts from machine contention, and a *different* set each run — `codex-shim`, then `combo-management-api`, then `codex-retained-root-serialization`, with individual durations reaching 260s. All three pass in isolation in seconds (69/69, 25/25, 6/6). The runner's own queue comment documents this failure mode: concurrent suites on one machine turn a ~210s run into a 13-minute one. None of those files touch the chat passthrough.
- Focused verification of the five directly-relevant suites: 206 pass / 0 fail.
- `bun x tsc --noEmit` clean.

Part of #1886.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Review readiness checklist

This PR stays in draft until every box below is ticked. Tick all four boxes once the requirements are met:

- [x] All CI tests are green on my local testing.
- [x] I pushed my PR to the latest dev commit.
- [x] I resolved all correct Codex and CodeRabbit findings.

- [x] My PR is ready for review.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved handling of Chat service tiers across standard and passthrough requests.
  * Unsupported or unauthorized tiers are now removed, while valid Fast tiers are forwarded consistently.
  * Native Chat requests and key-failover retries now apply the same service-tier rules as regular Chat requests.

* **Tests**
  * Added coverage for service-tier authorization, forwarding, fail-closed behavior, forced Fast mode, retries, and consistency across request paths.

### CODE REVIEW

**Comment by olddonkey** [comment]
Gate evidence at the merged head (`bdd114dc0`, dev tip `0cd85fbcd` merged in): **13,341 pass / 10 skip / 5 fail**.

Every failure was checked individually rather than assumed, and none implicates this diff:

| Failure | Verdict |
| --- | --- |
| `key-login-live-update` — modelCosts overlay | The pre-existing dev-side regression flagged on #1893/#1904/#1956/#2072. Reproduces on pristine `dev`. |
| `usage-log` — byte-prefix truncation (115s) | Timeout under load. Isolated: 35/35 with `lab-automation-final-coderabbit-regressions`, 3.8s total. |
| `codex-shim` — child invocation (2.0s) | Timeout under load. Isolated: 69/69. |
| `lab-automation-final-coderabbit-regressions` — CL-08 backoff (7.6s) | Timeout under load. Isolated: passes (above). |
| `cli-headless-parity` — combo set (81ms) | The one non-timeout shape, so worth naming: the spawned CLI returned exit 4 instead of 0 — a subprocess that failed to start under contention, not an assertion about behavior. Isolated: 23/23 in 156ms. |

This machine is currently running other agents' workloads alongside the suite, and `scripts/test.ts` only serializes *full-suite* runs against each other — its own comment records the failure mode ("four concurrent suites drove load average to 10 and turned a ~210s suite into a 13-minute one"). Across four runs of this branch the flake set was different every time (`codex-shim` → `combo-management-api` → `codex-retained-root-serialization` → this set), while the constant failure stayed the pre-existing dev one.

Focused verification of the suites this change actually touches: **206 pass / 0 fail** (`openai-chat-native-policy`, `fastwire-characterization-wire`, `openrouter-provider-routing`, `chat-completions-endpoint`, `openai-chat-hardening`), plus `bun x tsc --noEmit` clean.

**Comment by lidge-jun** [comment]
Independent review before merge found two correctness blockers, so this is held as needs-work now that #2042 (c472ad0f3) is on `dev`:

1. **Contradicts the exact-ID contract #2042 just landed.** The native-passthrough gate still matches `noStructuredOutputModels` through `modelInList` (prefix-before-`:` matching, src/types.ts:263), so `["gpt-oss"]` strips `response_format` from `gpt-oss:120b` — exactly the bleed #2042 removed with `includes(modelId)` (src/adapters/openai-chat.ts:123 on dev). Please rebase onto current dev and align the matching.

2. **Fast-policy parity is incomplete.** The shared helper consumes only `capability` from `ResolvedFastPolicy` and bypasses the `decideTier` branch that applies `fastWire.foreignCallerTiers === "drop"` (src/providers/fastwire.ts:338). With `supportsServiceTier:true + chatServiceTier:true + foreignCallerTiers:"drop"`, main Chat drops a foreign tier while the native passthrough forwards it; the parity table only exercises the default `verbatim` so the divergence is untested.

Also: no Cross-platform CI on the current head. Happy to re-review once rebased with both fixed.

**Comment by olddonkey** [comment]
Addressed the independent review in 01c823e24 after merging dev 63bfd149d. Native passthrough now uses exact noStructuredOutputModels matching and the shared tier decision semantics, including foreignCallerTiers drop; forced Fast uses the policy wire value and failover preserves the decision. Verification: focused native-chat suite 449 pass / 0 fail; typecheck, privacy scan, and diff check pass; full root suite 13,416 pass / 10 skip / 0 fail across 853 files. Please re-review the updated head.

File: tests/openai-chat-native-policy.test.ts [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+          status: 429,
+          headers: { "retry-after": "0" },
+        });
+      }
+      return Response.json({
+        id: "chatcmpl_native_tier_failover",
+        object: "chat.completion",
+        choices: [{ index: 0, message: { role: "assistant", content: "ok" }, finish_reason: "stop" }],
+      });
+    }) as typeof fetch;
```
[the remark this answers is not in the document]
Comment (olddonkey) [reply tests/openai-chat-native-policy.test.ts:212]: No code change needed: this file already captures originalFetch at module scope and its afterEach hook restores globalThis.fetch after every test, including both mocks referenced here and all failure paths. The same hook also clears provider cooldown state. Adding per-test restoration would duplicate the existing cleanup without changing behavior.

**Comment by lidge-jun** [comment]
## 리뷰 · 우선순위 64 / 80

#1886의 마지막 알려진 버그였다. native `/v1/chat/completions` 패스스루가 `chatServiceTier` 한 줄로 티어를 통과시켜, exact-model `false`와 `supportsServiceTier: false`를 무시했다. 두 Chat 면이 이제 같은 게이트를 탄다. hygiene 통과, `review-ready`, 체크리스트 4/4다. 메인테이너 hold(exact-ID 계약 #2042와 모순) 뒤에 저자가 `includes(modelId)`와 공유 `decideTier`로 고쳤다. `dev` FastWire와 맞아서 64다.

패스스루는 `src/adapters/openai-chat.ts`의 `buildOpenAIChatPassthroughRequest()`다. `decideTier(fastPolicy, fastMode, callerTier)`가 `set`이면 호출자 철자를 남기거나 forced Fast면 정책 와이어 값을 쓴다. `forward-caller`면 호출자 바이트를 그대로 둔다. 그 외는 스트립이다. `src/server/chat-native.ts`는 이 빌더에 `fastPolicyForModel(..., "chat")`과 `config.fastMode`를 넘기고, 키 로테이션 재빌드도 같은 클로저를 쓴다. 번역 Chat 경로는 `canSerializeOpenAIChatServiceTier()`로 같은 결정을 모은다.

의도적 차이는 남는다. Responses는 호출자 `"fast"`를 정규 와이어 값으로 바꾸고, 이 패스스루는 호출자 철자를 유지한다. 저자가 테스트로 그 분기를 고정했다. 남은 구멍도 명시했다. 패스스루는 `recordAdapterTier`를 안 불러서 스트립이 로그에 `callerTierDropped`로 안 보인다. 버그픽스 안에 관측 파이프 반쪽을 넣지 않은 선택은 맞다.

A0 characterization 셀이 “ignores exact-model false”에서 “honors exact-model false”로 뒤집혔다. `tests/openai-chat-native-policy.test.ts`가 fail-closed 두 단계, canonical vs foreign, unclassified, `"FAST"` 철자 유지, 두 Chat 경로 패리티를 본다. OpenRouter 라우팅 테스트는 시그니처만 맞추고 주장은 같다.

보안 서피스는 아니다. 티어 필드가 권한이 있을 때만 나간다. #2080/#2072와 같이 보면 이 게이트가 OpenRouter/xAI Fast의 바닥이다. 로그 공백은 후속 유닛으로 남겨도 된다.

## 해결방안

메인테이너가 업데이트된 head를 다시 읽고, #2042 exact-ID와 공유 게이트가 맞으면 머지하라. `callerTierDropped` 관측은 이 PR에 넣지 마라. 패리티 테스트를 유지하라. 번역 경로와 패스스루가 다시 갈라지는 게 이 버그의 재발이다.

이 댓글은 grok-bot이 작성했습니다

**Comment by lidge-jun** [comment]
Thanks for this, @olddonkey — closing as superseded by #2151, which carries **your change**, rebased onto current `dev`.

I owe you the same correction I owed @agentHits: I initially triaged this below my absorb threshold because it showed `CONFLICTING`, and treated that as a reason not to read further. A re-scoring pass on the actual diff put it at 67. The conflict was why it could not *merge*, not why it should score *low* — different questions, and I collapsed them. Rebasing turned out to be one import line.

The rebase, exactly: `src/adapters/openai-chat.ts` had an import-list conflict because `dev` added `AdapterTierMetadata` while your PR adds `decideTier` and `ResolvedFastPolicy`. Both sides kept, nothing dropped. Everything else applied cleanly.

Your characterization test is what makes this reviewable — flipping *"native chat passthrough honors exact-model false"* from a documented known-bug into a passing assertion means the contract is pinned rather than described. It is the test that fails when the source is reverted.

This is your second PR landing in this pass, alongside #2104 (OAuth Grok streaming, now #2147). Your work is credited in #2151's description.
